File: crates/human-vault/src/root_protection/capsule.rs
//! Root capsule: context-bound AES-GCM of the 32-byte vault root.
//! Mutable manifest revision is NEVER in capsule AAD (C04).
//! Encoding matches apps/pages protection/capsule.ts.

use aes_gcm::{
    aead::{Aead, KeyInit, Payload},
    Aes256Gcm, Nonce,
};
use base64::{engine::general_purpose::STANDARD, Engine};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use zeroize::Zeroize;

use super::canonicalize::canonicalize_to_bytes;
use super::error::ProtectionError;
use super::limits::{DOMAIN_CAPSULE, ROOT_KEY_BYTES};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProtectionContext {
    pub vault_id: String,
    pub root_key_id: String,
    pub root_epoch: u64,
    pub protector_id: String,
    pub purpose: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SealedBlobV1 {
    pub iv_b64: String,
    pub ct_b64: String,
}

fn context_aad(context: &ProtectionContext) -> Result<Vec<u8>, ProtectionError> {
    let value = json!({
        "domain": std::str::from_utf8(DOMAIN_CAPSULE).map_err(|_| ProtectionError::Crypto)?,
        "context": {
            "vaultId": context.vault_id,
            "rootKeyId": context.root_key_id,
            "rootEpoch": context.root_epoch,
            "protectorId": context.protector_id,
            "purpose": context.purpose,
        }
    });
    canonicalize_to_bytes(&value)
}

fn plaintext_bytes(
    context: &ProtectionContext,
    root_key: &[u8],
) -> Result<Vec<u8>, ProtectionError> {
    let value = json!({
        "v": 1,
        "domain": std::str::from_utf8(DOMAIN_CAPSULE).map_err(|_| ProtectionError::Crypto)?,
        "context": {
            "vaultId": context.vault_id,
            "rootKeyId": context.root_key_id,
            "rootEpoch": context.root_epoch,
            "protectorId": context.protector_id,
            "purpose": context.purpose,
        },
        "rootKeyB64": STANDARD.encode(root_key),
    });
    canonicalize_to_bytes(&value)
}

/// Seal a 32-byte root under `kek` with context AAD.
///
/// # Errors
///
/// Returns typed failures for length, encoding, or AEAD errors.
pub fn seal_root_capsule(
    kek: &[u8; 32],
    context: &ProtectionContext,
    root_key: &[u8],
    iv: &[u8; 12],
) -> Result<SealedBlobV1, ProtectionError> {
    if root_key.len() != ROOT_KEY_BYTES {
        return Err(ProtectionError::InvalidKeyLength);
    }
    let aad = context_aad(context)?;
    let mut pt = plaintext_bytes(context, root_key)?;
    let cipher = Aes256Gcm::new_from_slice(kek).map_err(|_| ProtectionError::Crypto)?;
    let ct = cipher
        .encrypt(
            Nonce::from_slice(iv),
            Payload {
                msg: &pt,
                aad: &aad,
            },
        )
        .map_err(|_| ProtectionError::Crypto)?;
    pt.zeroize();
    Ok(SealedBlobV1 {
        iv_b64: STANDARD.encode(iv),
        ct_b64: STANDARD.encode(ct),
    })
}

/// Open a sealed capsule, requiring the independently chosen context.
///
/// # Errors
///
/// Returns typed failures for nonce, AEAD, context, or key length.
pub fn open_root_capsule(
    kek: &[u8; 32],
    expected: &ProtectionContext,
    sealed: &SealedBlobV1,
) -> Result<[u8; 32], ProtectionError> {
    let iv = STANDARD
        .decode(&sealed.iv_b64)
        .map_err(|_| ProtectionError::MalformedEncoding("capsule iv".into()))?;
    if iv.len() != 12 {
        return Err(ProtectionError::MalformedEncoding(
            "capsule iv length".into(),
        ));
    }
    let ct = STANDARD
        .decode(&sealed.ct_b64)
        .map_err(|_| ProtectionError::MalformedEncoding("capsule ct".into()))?;
    let aad = context_aad(expected)?;
    let cipher = Aes256Gcm::new_from_slice(kek).map_err(|_| ProtectionError::Crypto)?;
    let raw = cipher
        .decrypt(
            Nonce::from_slice(&iv),
            Payload {
                msg: &ct,
                aad: &aad,
            },
        )
        .map_err(|_| ProtectionError::CapsuleAuthFailed)?;
    let value: Value = serde_json::from_slice(&raw)
        .map_err(|e| ProtectionError::MalformedEncoding(e.to_string()))?;
    let v = value
        .get("v")
        .and_then(Value::as_u64)
        .ok_or(ProtectionError::UnsupportedVersion(0))?;
    if v != 1 {
        return Err(ProtectionError::UnsupportedVersion(v as u32));
    }
    let domain = value
        .get("domain")
        .and_then(Value::as_str)
        .ok_or_else(|| ProtectionError::MalformedEncoding("domain".into()))?;
    if domain.as_bytes() != DOMAIN_CAPSULE {
        return Err(ProtectionError::UnsupportedVersion(0));
    }
    let ctx = value
        .get("context")
        .ok_or_else(|| ProtectionError::MalformedEncoding("context".into()))?;
    let parsed = ProtectionContext {
        vault_id: ctx
            .get("vaultId")
            .and_then(Value::as_str)
            .ok_or_else(|| ProtectionError::MalformedEncoding("vaultId".into()))?
            .to_string(),
        root_key_id: ctx
            .get("rootKeyId")
            .and_then(Value::as_str)
            .ok_or_else(|| ProtectionError::MalformedEncoding("rootKeyId".into()))?
            .to_string(),
        root_epoch: ctx
            .get("rootEpoch")
            .and_then(Value::as_u64)
            .ok_or_else(|| ProtectionError::MalformedEncoding("rootEpoch".into()))?,
        protector_id: ctx
            .get("protectorId")
            .and_then(Value::as_str)
            .ok_or_else(|| ProtectionError::MalformedEncoding("protectorId".into()))?
            .to_string(),
        purpose: ctx
            .get("purpose")
            .and_then(Value::as_str)
            .ok_or_else(|| ProtectionError::MalformedEncoding("purpose".into()))?
            .to_string(),
    };
    if &parsed != expected {
        return Err(ProtectionError::ContextMismatch);
    }
    let root_b64 = value
        .get("rootKeyB64")
        .and_then(Value::as_str)
        .ok_or_else(|| ProtectionError::MalformedEncoding("rootKeyB64".into()))?;
    let root = STANDARD
        .decode(root_b64)
        .map_err(|_| ProtectionError::InvalidKeyLength)?;
    let key: [u8; 32] = root
        .try_into()
        .map_err(|_| ProtectionError::InvalidKeyLength)?;
    Ok(key)
}

File: packages/app-core/src/lib/vault/protection/capsule.ts
/**
 * Root capsule: context-bound AES-GCM package of the 32-byte vault root.
 * Mutable manifest revision is NEVER in capsule AAD (C04).
 */

import {
  type BoundaryValue,
  isJsonObject,
  isNumber,
  isString,
  overlapCast,
} from "@opensesame/os-domain";
import {
  DOMAIN_CAPSULE,
  type ProtectionContext,
  ROOT_KEY_BYTES,
  type SealedBlobV1,
} from "@opensesame/vault-core";
import { b64ToBytes, bytesToB64 } from "@opensesame/vault-core";
import { canonicalizeToBytes } from "./canonicalize.js";
import { ProtectionError } from "./errors.js";

export type RootCapsulePlaintext = {
  v: 1;
  domain: typeof DOMAIN_CAPSULE;
  context: ProtectionContext;
  rootKeyB64: string;
};

function contextAad(context: ProtectionContext): Uint8Array {
  return canonicalizeToBytes({
    domain: DOMAIN_CAPSULE,
    context,
  });
}

export async function sealRootCapsule(
  kek: CryptoKey,
  context: ProtectionContext,
  rootKey: Uint8Array,
  /** Fixed IV for cross-language vectors only. */
  ivOverride?: Uint8Array,
): Promise<SealedBlobV1> {
  if (rootKey.byteLength !== ROOT_KEY_BYTES) {
    throw new ProtectionError(
      "invalid_key_length",
      `Root key must be ${ROOT_KEY_BYTES} bytes.`,
    );
  }
  if (ivOverride !== undefined && ivOverride.byteLength !== 12) {
    throw new ProtectionError("invalid_nonce", "Capsule IV must be 12 bytes.");
  }
  const plaintext: RootCapsulePlaintext = {
    v: 1,
    domain: DOMAIN_CAPSULE,
    context,
    rootKeyB64: bytesToB64(rootKey),
  };
  const iv = ivOverride ?? crypto.getRandomValues(new Uint8Array(12));
  const ct = await crypto.subtle.encrypt(
    {
      name: "AES-GCM",
      iv,
      additionalData: overlapCast(contextAad(context)),
    },
    kek,
    canonicalizeToBytes(plaintext),
  );
  return { ivB64: bytesToB64(iv), ctB64: bytesToB64(new Uint8Array(ct)) };
}

function parseRootCapsulePlaintext(raw: ArrayBuffer): RootCapsulePlaintext {
  const decoded: BoundaryValue = JSON.parse(new TextDecoder().decode(raw));
  if (!isJsonObject(decoded)) {
    throw new ProtectionError(
      "malformed_encoding",
      "Root capsule plaintext is not an object.",
    );
  }
  if (decoded.v !== 1 || decoded.domain !== DOMAIN_CAPSULE) {
    throw new ProtectionError(
      "unsupported_version",
      "Unsupported root capsule version.",
    );
  }
  if (!isString(decoded.rootKeyB64)) {
    throw new ProtectionError(
      "malformed_encoding",
      "Root capsule rootKeyB64 is malformed.",
    );
  }
  if (!isJsonObject(decoded.context)) {
    throw new ProtectionError(
      "malformed_encoding",
      "Root capsule context is malformed.",
    );
  }
  const ctx = decoded.context;
  if (
    !isString(ctx.vaultId) ||
    !isString(ctx.rootKeyId) ||
    !isNumber(ctx.rootEpoch) ||
    !isString(ctx.protectorId) ||
    (ctx.purpose !== "human-vault-root" && ctx.purpose !== "workload-root")
  ) {
    throw new ProtectionError(
      "malformed_encoding",
      "Root capsule context fields are malformed.",
    );
  }
  return {
    v: 1,
    domain: DOMAIN_CAPSULE,
    context: {
      vaultId: ctx.vaultId,
      rootKeyId: ctx.rootKeyId,
      rootEpoch: ctx.rootEpoch,
      protectorId: ctx.protectorId,
      purpose: ctx.purpose,
    },
    rootKeyB64: decoded.rootKeyB64,
  };
}

export async function openRootCapsule(
  kek: CryptoKey,
  expected: ProtectionContext,
  sealed: SealedBlobV1,
): Promise<Uint8Array> {
  const iv = b64ToBytes(sealed.ivB64);
  if (iv.byteLength !== 12) {
    throw new ProtectionError("invalid_nonce", "Capsule IV must be 12 bytes.");
  }
  let raw: ArrayBuffer;
  try {
    raw = await crypto.subtle.decrypt(
      {
        name: "AES-GCM",
        iv: overlapCast(iv),
        additionalData: overlapCast(contextAad(expected)),
      },
      kek,
      overlapCast(b64ToBytes(sealed.ctB64)),
    );
  } catch {
    throw new ProtectionError(
      "capsule_auth_failed",
      "Root capsule authentication failed.",
    );
  }
  const parsed = parseRootCapsulePlaintext(raw);
  if (!contextsEqual(parsed.context, expected)) {
    throw new ProtectionError(
      "context_mismatch",
      "Root capsule context does not match expected protection context.",
    );
  }
  const rootKey = b64ToBytes(parsed.rootKeyB64);
  if (rootKey.byteLength !== ROOT_KEY_BYTES) {
    throw new ProtectionError(
      "invalid_key_length",
      `Recovered root key must be ${ROOT_KEY_BYTES} bytes.`,
    );
  }
  return rootKey;
}

export function contextsEqual(
  left: ProtectionContext,
  right: ProtectionContext,
): boolean {
  return (
    left.vaultId === right.vaultId &&
    left.rootKeyId === right.rootKeyId &&
    left.rootEpoch === right.rootEpoch &&
    left.protectorId === right.protectorId &&
    left.purpose === right.purpose
  );
}
